### PR TITLE
adds DnD support to TextColumnField component

### PR DESCRIPTION
**Pull request by invincibleJai**
**Fixes**: 
https://issues.redhat.com/browse/ODC-5635

**Solution Description**: 
Update TextColumnField component to support DragNdrop functionality based on props.

**Screen shots / Gifs for design review**: 

image

Apr-20-2021 13-03-21-Text-col-field

these are demonstration, actual usage of this will happen with edit form work.

@openshift/team-devconsole-ux

**Test setup:**
pass prop `dndEnabled` to any form with `TextColumnField`.

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    },
+  });
+
+  return (
+    <div ref={(node) => preview(drop(node))} style={{ opacity }}>
+      <Flex style={{ marginBottom: 'var(--pf-global--spacer--sm)' }}>
+        {!!dndEnabled && (
+          <FlexItem style={{ cursor: 'move' }}>
+            <div ref={drag}>
+              <GripLinesVerticalIcon />
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx:59]: The icon seems to be different in the designs

Screenshot 2021-04-20 at 2 00 56 PM

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      }
+      // monitor item updated here to avoid expensive index searches.
+      item.idx = idx;
+    },
+  });
+
+  return (
+    <div ref={(node) => preview(drop(node))} style={{ opacity }}>
+      <Flex style={{ marginBottom: 'var(--pf-global--spacer--sm)' }}>
+        {!!dndEnabled && (
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx:56]:
```suggestion
        {dndEnabled && (
```

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                  );
+                })}
+              </FormGroup>
+            </DndProvider>
+            {!isReadOnly && (
+              <MultiColumnFieldFooter
+                addLabel={addLabel}
+                onAdd={() => {
+                  const newValue = '';
+                  arrayHelpers.push(newValue);
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx:51]:
```suggestion
                  arrayHelpers.push('');
```

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                  onChange && onChange([...rowValues, newValue]);
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx:52]:
```suggestion
                  onChange && onChange([...rowValues, '']);
```

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  const [field, { touched, error }] = useField<string[]>(name);
+  useFormikValidationFix(field.value);
+  const rowValues = field.value ?? [];
+  const fieldId = getFieldId(name, 'single-column');
+  const isValid = !(touched && error);
+  return (
+    <FieldArray
+      name={name}
+      render={(arrayHelpers) => {
+        return (
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx:23]:
```suggestion
      render={(arrayHelpers) => (
```

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          <>
+            <DndProvider backend={HTML5Backend}>
+              <FormGroup
+                fieldId={fieldId}
+                label={label}
+                validated={isValid ? 'default' : 'error'}
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx:29]: you can import the const `ValidatedOptions` from `@patternfly/react-core`.

```suggestion
                validated={isValid ? ValidatedOptions.default : ValidatedOptions.error}
```

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  idx,
+  name,
+  tooltip,
+  placeholder,
+  isReadOnly,
+  onChange,
+  rowValues,
+  disableDeleteRow,
+  arrayHelpers,
+  dndEnabled,
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx:27]:
```suggestion
  dndEnabled = false,
```

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx (hunk 1 above)
Comment (invincibleJai) [reply frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx:59]: check this one in the design doc

image

confirmed with design team have to use `GripVertical` cc @divyanshiGupta

Comment (divyanshiGupta) [reply frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnItem.tsx:59]: Yes, the dnd icon will be replaced with pf icon across console so here we need to use the pf one too.

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx (hunk 3 above)
Comment (invincibleJai) [reply frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx:51]: Ah, this is from old code I just did refactor but will get this updated.

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx (hunk 4 above)
Comment (invincibleJai) [reply frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx]: Ah, this is from the old code I just did refactor but will get this updated.

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx (hunk 5 above)
Comment (invincibleJai) [reply frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx]: Ah, this is from old code I just did refactor but will get this updated.

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+              helperText={helpText}
+            >
+              {rowValues.map((v, idx) => {
+                return (
+                  <TextColumnItem
+                    key={`${idx.toString()}`}
+                    idx={idx}
+                    arrayHelpers={arrayHelpers}
+                    rowValues={rowValues}
+                    {...props}
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/TextColumnField.tsx:39]: shouldn't this be 
```suggestion
                  <TextColumnItem
                   {...props}
                    key={`${idx.toString()}`}
                    idx={idx}
                    arrayHelpers={arrayHelpers}
                    rowValues={rowValues}
                   
```

File: frontend/packages/console-shared/src/components/formik-fields/text-column-field/__tests__/TextColumnItem.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    const wrapper = shallow(
+      <TextColumnItem
+        name={'fieldName'}
+        label={'label value'}
+        idx={0}
+        rowValues={['']}
+        arrayHelpers={mockArrayHelper}
+      />,
+    );
+    expect(wrapper.exists()).toBe(true);
```
Comment (sahil143) [inline frontend/packages/console-shared/src/components/formik-fields/text-column-field/__tests__/TextColumnItem.spec.tsx:53]:
```suggestion
    expect(wrapper.isEmptyRender()).toBe(false);
```

**Review by sahil143 — COMMENTED** [review]
@invincibleJai I don't think we should show the drag handle if there is only one input field.

Screenshot 2021-04-20 at 5 30 48 PM

**Comment by invincibleJai** [comment]
> @invincibleJai I don't think we should show the drag handle if there is only one input field.
> 
> Screenshot 2021-04-20 at 5 30 48 PM

did you mean drag icon, there is no harm in showing it. Let me check with UXD @bmignano WDYT?

**Comment by bmignano** [comment]
> > @invincibleJai I don't think we should show the drag handle if there is only one input field.
> > Screenshot 2021-04-20 at 5 30 48 PM
> 
> did you mean drag icon, there is no harm in showing it. Let me check with UXD as the design had it too @bmignano WDYT?

I agree with @invincibleJai that there is no harm in showing it! It is also aligned with what we do today – even if there's only one row we still show it. So, I would recommend just updating the icon and not changing this behavior.

**Review by sahil143 — APPROVED** [review]
/lgtm

verified changes locally

**Comment by christianvogt** [comment]
Tested and working as expected.
/label qe-approved

No other acks necessary as this is a dev PR adding support in a reusable component. 

/approve

**Pull request merged**